### PR TITLE
add documentation about new simplify connector form fields

### LINKED ISSUE #23531: Document new spec behavior
## Context

Some new features are being added to connector specs, as defined in the [Simplify Connector Configuration Tech Spec](https://docs.google.com/document/d/1FleXuPvAaGKrrsFBWsJdD0kRg_UF4Ea17qj6JgqsTF8).

These features should be clearly documented so that it is easy for connector developers to understand what features are available to them when writing connector specs.

## Acceptance Criteria
- The [Connector Specification Reference](https://docs.airbyte.com/connector-development/connector-specification-reference/) clearly document how the following connector spec features are used, and what the default behavior is if they are not used, and examples with screenshots etc.:
  - "order"
  - "group"
  - "alwaysShow"
  - "patternDescriptor"
- The docs clearly explain what is NOT allowed in connector specs, both existing validations and the new validations added in https://github.com/airbytehq/airbyte/issues/23529

### PR DESCRIPTION
**Pull request by lmossman**
## What
Resolves https://github.com/airbytehq/airbyte/issues/23531

Adds documentation about the new fields introduced in the [Simplify Connector Configuration project](https://github.com/airbytehq/airbyte/issues/23526)

## How
Adds descriptions, examples, and screenshots to the docs to explain the new fields, as well as the rules about them.

**Note**: since these new fields are still hidden behind a feature flag, we should wait to merge in these documentation changes until after the feature flag has been removed.

### CODE REVIEW

File: docs/connector-development/connector-specification-reference.md [hunk 1]
```diff
@@ -43,17 +41,132 @@ For example, using the following spec:
...
-will result in the following configuration on the UI: 
+will result in the following configuration on the UI:
 
 ![Screen Shot 2021-11-18 at 7 14 04 PM](https://user-images.githubusercontent.com/6246757/142558797-135f6c73-f05d-479f-9d88-e20cae85870c.png)
 
-
 :::info
 
-Within an object definition, if some fields have the `order` property defined, and others don't, then the fields without the `order` property defined should be rendered last in the UI. Among those elements (which don't have `order` defined), no ordering is guaranteed. 
+Within an object definition, if some fields have the `order` property defined, and others don't, then the fields without the `order` property defined should be rendered last in the UI. Among those elements (which don't have `order` defined), no ordering is guaranteed.
```
Comment (flash1293) [inline docs/connector-development/connector-specification-reference.md:50]: Should we add a note to the "order" section about duplicated order as well? Seems relevant in both contexts

File: docs/connector-development/connector-specification-reference.md [hunk 2]
```diff
@@ -43,17 +41,132 @@ For example, using the following spec:
...
 
 :::
 
+### Collapsing optional fields
+
+By default, all optional fields will be collapsed into an `Optional fields` section which can be expanded or collapsed by the user. This helps streamline the UI for setting up a connector by initially focusing attention on the required fields only.
```
Comment (flash1293) [inline docs/connector-development/connector-specification-reference.md:56]: should we add a few words here that optional fields are moved to the bottom of the form/group if they don't have an explicit order specified?

**Review by flash1293 — COMMENTED** [review]
Looks mostly good, left some remarks

File: docs/.gitbook/assets/connector-form-groups.png
Comment (sherifnada) [inline docs/.gitbook/assets/connector-form-groups.png:1]: fwiw you don't need to put the files in `.gitbook` you can just put them in the same dir as these docs, for ease of use

File: docs/connector-development/connector-specification-reference.md [hunk 3]
```diff
@@ -43,17 +41,138 @@ For example, using the following spec:
...
 ```
 
-will result in the following configuration on the UI: 
+will result in the following configuration on the UI:
 
 ![Screen Shot 2021-11-18 at 7 14 04 PM](https://user-images.githubusercontent.com/6246757/142558797-135f6c73-f05d-479f-9d88-e20cae85870c.png)
 
+:::info
+
+Within an object definition, if some fields have the `order` property defined, and others don't, then the fields without the `order` property defined should be rendered last in the UI. Among those elements (which don't have `order` defined), no ordering is guaranteed.
```
Comment (sherifnada) [inline docs/connector-development/connector-specification-reference.md:50]:
```suggestion
Within an object definition, if some fields have the `order` property defined, and others don't, then the fields without the `order` property defined will be rendered last in the UI. Among those elements (which don't have `order` defined), no ordering is guaranteed.
```

**Review by sherifnada — APPROVED** [review]
Looks great!

cc @evantahler

**Review by flash1293 — APPROVED** [review]
LGTM

File: docs/connector-development/connector-specification-reference.md [hunk 4]
```diff
@@ -105,6 +225,26 @@ Results in the following form:
...
+The `pattern_descriptor` property allows the connector developer to set a human-readable format that should be displayed above the field, and if set in conjunction with a `pattern`, this `pattern_descriptor` will be used in the invalid format error message instead of the raw regex.
+
+For example, having a field in the spec like:
+
+```
+"start_date": {
+  "type": "string",
+  "title": "Start date",
+  "pattern": "^[0-9]{4}-[0-9]{2}-[0-9]{2}$",
+  "pattern_descriptor": "YYYY-MM-DD"
```
Comment (evantahler) [inline docs/connector-development/connector-specification-reference.md:241]: This is nice.

File: docs/connector-development/connector-specification-reference.md [hunk 5]
```diff
@@ -83,9 +202,10 @@ By default, string inputs in the UI can lose their linebreaks. In order to accep
...
 
 ### Hiding inputs in the UI
-In some rare cases, a connector may wish to expose an input that is not available in the UI, but is still potentially configurable when running the connector outside of Airbyte, or via the UI. For example, exposing a very technical configuration like the page size of an outgoing HTTP requests may only be relevant to power users, and therefore shouldn't be available via the UI but might make sense to expose via the API. 
 
-In this case, use the `"airbyte_hidden": true` keyword to hide that field from the UI. E.g: 
+In some rare cases, a connector may wish to expose an input that is not available in the UI, but is still potentially configurable when running the connector outside of Airbyte, or via the UI. For example, exposing a very technical configuration like the page size of an outgoing HTTP requests may only be relevant to power users, and therefore shouldn't be available via the UI but might make sense to expose via the API.
+
+In this case, use the `"airbyte_hidden": true` keyword to hide that field from the UI. E.g:
```
Comment (evantahler) [inline docs/connector-development/connector-specification-reference.md:208]: I had no idea!

File: docs/connector-development/connector-specification-reference.md [hunk 6]
```diff
@@ -43,17 +41,138 @@ For example, using the following spec:
...
 ```
 
-will result in the following configuration on the UI: 
+will result in the following configuration on the UI:
 
 ![Screen Shot 2021-11-18 at 7 14 04 PM](https://user-images.githubusercontent.com/6246757/142558797-135f6c73-f05d-479f-9d88-e20cae85870c.png)
 
+:::info
+
+Within an object definition, if some fields have the `order` property defined, and others don't, then the fields without the `order` property defined will be rendered last in the UI. Among those elements (which don't have `order` defined), no ordering is guaranteed.
```
Comment (evantahler) [inline docs/connector-development/connector-specification-reference.md:50]: 👍 nice fallback

**Review by evantahler — APPROVED** [review]

**Comment by flash1293** [comment]
@lmossman this looks ready, can we merge it?

**Comment by flash1293** [comment]
Makes sense @lmossman , just making sure

**Pull request merged**